expr/plan.go

```diff
@@ -358,3 +360,28 @@ func (p *Plan) Run(dataMap DataMap) ([]models.Series, error) {
 func (p Plan) Clean() {
 	p.dataMap.Clean()
 }
+
+func (p Plan) CheckedClean(targets []string) bool {
+	type ReqSeries struct {
+		req    Req
+		series models.Series
+	}
+	// Map of pointer to end of slice to the first series we found with that address
+	addrs := make(map[*schema.Point]ReqSeries)
+	for req, series := range p.dataMap {
+		for _, serie := range series {
+			dpCap := cap(serie.Datapoints)
+			if dpCap > 0 {
+				addr := &(serie.Datapoints[0:dpCap][dpCap-1])
+				if val, ok := addrs[addr]; ok {
+					log.Errorf("Found results sharing a slice: query = %v, req1 = %v, series = %v, req2 = %v, series = %v", targets, val.req, val.series, req, serie)
+					// Don't want to spew errors so move on (don't clean this one)
+					return false
```

Comment:
interesting choice to do this on the plan level, where, upon hitting this issue, we abort cleaning the entire plan/datamap.

if this logic was implemented instead in the datamap, we could strip out the double ones but still reclaim the majority of pointslices

Reply:
Yeah, I originally did continue and clean as much as possible. But at any point I see double slices here, I do start to wonder if it's possible that the issue comes from before the plan and the plan is a victim (having grabbed the same slice twice, not duplicated it). So, at that point, the pointslicepool is possibly corrupted and it's better to *not* return things to it (because some of the slices might already be in there!!).

Reply:
ok that's reasonable. basically when things go wrong, assume the worst.
this way hitting a small bug may result in significant alloc rate /GC impact, all the more motivation for us to fix these bugs.
maybe something to document somewhere, though not sure where.